### PR TITLE
feat: Improve the "logs" experience and reduce number of backend calls

### LINKED ISSUE #852: Stop fetching of logs on finished pipelines
The logs of a finished pipeline are still loaded from the backend every few seconds. This produces unnecessary load on the backend and disappearing logs in the frontend. On finished pipelines, the logs and events should only be loaded once.
Also, loading of logs can take some time, especially for older pipelines. We should have a loading animation.

### LINKED ISSUE #864: Pipeline deletion fails with existing job runs
If a pipeline has manually triggered runs, the deletion fails. On pipeline deletion, we should add the cascading delete flag in the database.

### PR DESCRIPTION
**Pull request by MoritzWeber0**
Please check the commit messages for detailed information.

The main improvements are: 

- Logs and events for finished jobs are only fetched once, not every two seconds (they won't change anymore)
- The query time range for Grafana Loki calls was reduced, leading to less calls inside of Grafana Loki
- Add handling of TooManyRequests error in Grafana Loki 
- Jobs are properly terminated. Logs and events are fetched before termination. 

Resolves #852
Resolves #864

### CODE REVIEW

File: frontend/src/app/projects/models/backup-settings/view-logs-dialog/view-logs-dialog.component.ts [hunk 1]
```diff
@@ -29,26 +29,31 @@ export class ViewLogsDialogComponent {
...
       this.pipelineRunService.pipelineRun$.pipe(filter(Boolean)),
     ])
       .pipe(
         untilDestroyed(this),
-        switchMap(([project, model, pipeline, pipelineRun]) =>
-          this.pipelineRunService.getEvents(
+        filter(
+          ([_project, _model, _pipeline, pipelineRun]) =>
+            this.logs === undefined ||
+            !this.pipelineRunService.pipelineRunIsFinished(pipelineRun.status)
```
Comment (dominik003) [inline frontend/src/app/projects/models/backup-settings/view-logs-dialog/view-logs-dialog.component.ts:47]:
```suggestion
            !(this.logs && this.pipelineRunService.pipelineRunIsFinished(pipelineRun.status))
```

File: frontend/src/app/projects/models/backup-settings/view-logs-dialog/view-logs-dialog.component.ts [hunk 2]
```diff
@@ -29,26 +29,31 @@ export class ViewLogsDialogComponent {
...
             model.slug,
             pipeline.id,
             pipelineRun.id
-          )
-        )
+          );
+        })
       )
       .subscribe({
         next: (res: string) => {
```
Comment (dominik003) [inline frontend/src/app/projects/models/backup-settings/view-logs-dialog/view-logs-dialog.component.ts:59]:
```
next: (res: string) => (this.events = res),
error: () => (this.events = "Couldn't fetch events"),
```
Again nothing that you directly changed but for these one line assignments at least I prefer to just do it like this. Whats your opinion on that?

File: frontend/src/app/projects/models/backup-settings/view-logs-dialog/view-logs-dialog.component.ts [hunk 3]
```diff
@@ -69,6 +74,11 @@ export class ViewLogsDialogComponent {
     ])
       .pipe(
         untilDestroyed(this),
+        filter(
+          ([_project, _model, _pipeline, pipelineRun]) =>
+            this.logs === undefined ||
+            !this.pipelineRunService.pipelineRunIsFinished(pipelineRun.status)
```
Comment (dominik003) [inline frontend/src/app/projects/models/backup-settings/view-logs-dialog/view-logs-dialog.component.ts:80]:
```suggestion
            !(this.logs && this.pipelineRunService.pipelineRunIsFinished(pipelineRun.status))
```

**Review by dominik003 — COMMENTED** [review]
Nice improvements! Really like the new exception handling to reduce the number of repetitive fastapi exceptions.

File: backend/capellacollab/core/logging/loki.py [hunk 4]
```diff
@@ -81,8 +84,12 @@ def fetch_logs_from_loki(
                 PROMTAIL_CONFIGURATION["lokiUsername"],
                 PROMTAIL_CONFIGURATION["lokiPassword"],
             ),
-            timeout=10,
+            timeout=5,
         )
+
+        if response.status_code == 429:
```
[the remark this answers is not in the document]
Comment (MoritzWeber0) [reply backend/capellacollab/core/logging/loki.py:90]: This is out of the `fastapi` content. Instead, I changes it to compare the `HTTPStatus`: 
```
if response.status_code == http.HTTPStatus.TOO_MANY_REQUESTS:
```

File: backend/capellacollab/projects/toolmodels/backups/runs/interface.py [hunk 5]
```diff
@@ -200,11 +202,34 @@ def _search_and_kill_timed_out_jobs(run: models.DatabasePipelineRun):
...
+
+def _terminate_job(
+    run: models.DatabasePipelineRun, updated_status: models.PipelineRunStatus
+):
+    run.status = updated_status
+    run.end_time = datetime.datetime.now()
+    # Fetch logs and events one last time before killing
+    try:
+        _fetch_logs_of_jobs(run)
+        _fetch_events_of_jobs(run)
```
[the remark this answers is not in the document]
Comment (MoritzWeber0) [reply backend/capellacollab/projects/toolmodels/backups/runs/interface.py:216]: The `end_time` should indicate when the pipeline finished, not when the last logs were fetched. It's a totally valid situation that the logs are fetched after the pipeline has ended (with the current code that happens each time). And this is exactly what we want, fetch the logs after the pipeline is finished.

File: backend/capellacollab/projects/toolmodels/backups/runs/interface.py [hunk 6]
```diff
@@ -250,11 +275,24 @@ def _update_status_of_job(
         run.status = models.PipelineRunStatus.UNKNOWN
         return
     current_status = _map_k8s_to_internal_status(job)
+
+    if _job_is_finished(current_status):
+        _terminate_job(run, current_status)
+        return
```
[the remark this answers is not in the document]
Comment (MoritzWeber0) [reply backend/capellacollab/projects/toolmodels/backups/runs/interface.py:281]: A job can terminate with a specific status (e.g., SUCCESS / FAILURE / TIMEOUT). So I don't get the point why `_terminate_job` should not get an updated_status.

[the remark this answers is not in the document]
Comment (MoritzWeber0) [reply backend/capellacollab/projects/toolmodels/backups/runs/interface.py]: Implemented option 1.

File: backend/capellacollab/projects/toolmodels/backups/runs/routes.py [hunk 7]
```diff
@@ -118,6 +118,18 @@ def get_events(
     )
 
 
+def _determine_end_time_from_pipeline_run(
+    pipeline_run: models.DatabasePipelineRun,
+) -> datetime.datetime:
+    max_pipeline_run_duration = datetime.timedelta(
+        minutes=65
```
[the remark this answers is not in the document]
Comment (MoritzWeber0) [reply backend/capellacollab/projects/toolmodels/backups/runs/routes.py:125]: Is now added as configuration option `pipelines.timeout`.

File: frontend/src/app/projects/models/backup-settings/pipeline-runs/wrapper/pipeline-run-wrapper/pipeline-run-wrapper.component.ts [hunk 8]
```diff
@@ -41,10 +41,14 @@ export class PipelineRunWrapperComponent implements OnDestroy {
       this.pipelineRunService.resetPipelineRun();
     });
 
-    this.updatePipelineRun();
```
[the remark this answers is not in the document]
Comment (MoritzWeber0) [reply frontend/src/app/projects/models/backup-settings/pipeline-runs/wrapper/pipeline-run-wrapper/pipeline-run-wrapper.component.ts:44]: > When I looked into this here I just though about the `combineLatest` directly above here. Isn't it too complex and could be simplified quite a bit?
> 
> 1. `pipeline` can only be either set or undefined. Therefore simply using `filter(Boolean)` should be sufficient for the first filter or?

We had that in another comment already. `filter(Boolean)` is comparable to `filter((pipeline) => pipeline !== undefined)`. It's the opposite of what we want to approach here. 

> 2. we don't really need the result of the map here so we could just use `filter((params) => !!params?.pipelineRun)` for the second part

Changed it to `params?.pipelineRun === undefined`. Even though it doesn't make any difference, but I prefer explicit type checks over more general checks.

File: frontend/src/app/projects/models/backup-settings/view-logs-dialog/view-logs-dialog.component.ts (hunk 1 above)
Comment (MoritzWeber0) [reply frontend/src/app/projects/models/backup-settings/view-logs-dialog/view-logs-dialog.component.ts:47]: First of all, `this.logs === undefined` is need here. `this.logs` could also be an empty array, which has to be handled differently.

Also, your suggestion is different. There is no `!` before the `this.logs` in the origin implementation.

File: frontend/src/app/projects/models/backup-settings/view-logs-dialog/view-logs-dialog.component.ts (hunk 3 above)
Comment (MoritzWeber0) [reply frontend/src/app/projects/models/backup-settings/view-logs-dialog/view-logs-dialog.component.ts:80]: Same as above.

**Review by dominik003 — APPROVED** [review]

**Pull request merged**